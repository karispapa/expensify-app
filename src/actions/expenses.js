import uuid from 'uuid';


//ADD_EXPENSE
// action generator for add expense
export const addExpense = ({
  description = '',
  text = '',
  amount =0,
createdAt = 0}
  ={})=> ({
    type: 'ADD_EXPENSE',
    expense: {
      id: uuid(),
      description,
      text,
      amount,
      createdAt
    }
});

//REMOVE_EXPENSE
// action generator for remove expense
export const removeExpense = ({id} = {})=> ({
  type: 'REMOVE_EXPENSE',
  id
});
//EDIT_EXPENSE
// action generator for edit expense
export const editExpense = (id, updates)=> ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});