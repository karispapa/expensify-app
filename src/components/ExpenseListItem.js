import React from 'react';
import {Link} from 'react-router-dom'
// import { connect } from 'react-redux';
// import { removeExpense } from '../actions/expenses'

const ExportListItem = ({id, description, amount, createdAt})=>(

  <div>
    <Link to={`/edit/${id}`}>
     <h3>{description} </h3>
    </Link>
    
    <p> {createdAt} ksh: {amount} </p>
  </div>
);

export default ExportListItem; 