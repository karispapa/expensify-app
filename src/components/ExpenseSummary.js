import React from 'react';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';
import calculateTotals from '../selectors/expensesTotal';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral'


export const ExpenseSummary = ({expenseCount, expenseTotal}) => {
  
  const expenseWord = expenseCount === 1 ? 'expense': 'expenses'
  
    return(
      <div className="page-header">
        <div className="content-container">
          <div className="page-header__title">
          {expenseCount === 0 ? <h1>You have no expenses</h1>:  
            <h1 >Viewing <span>{expenseCount}</span>  {expenseWord} totaling <span>{expenseTotal}</span> </h1>
          }
         </div>
         <div className="page-header__actions">
          <Link className="button" to="/create">Add Expense</Link>
         </div>

        </div>
        
      
      </div>
    )
  }

const mapStateToProps = (state)=>{
  const expenses = selectExpenses(state.expenses, state.filters)
  return{
  expenseCount: expenses.length,
  expenseTotal: numeral(calculateTotals(expenses)).format('$0,0.00') 
}
};

export default connect(mapStateToProps)(ExpenseSummary)