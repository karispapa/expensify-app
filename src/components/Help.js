import React from 'react';

const Help = ()=> (
  <div>
    This is the Help Page  
  </div>
);

export default Help;
