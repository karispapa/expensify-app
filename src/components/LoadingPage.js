import React from 'react';

const LoadingPage = ()=>(
  <div className="loader">
    <img className="loader__image" src="/images/loader.gif"/>
    <span className="loader__text">Please Wait</span>
  </div>
)

export default LoadingPage