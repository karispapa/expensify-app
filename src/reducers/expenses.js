

const expensesReducerDefault = [];

// Expenses reducer

export default (state = expensesReducerDefault, action)=>{

  switch(action.type){

    case 'ADD_EXPENSE':
      return [...state, action.expense]

    case 'REMOVE_EXPENSE':
      return state.filter(({id})=> action.id !== id) // filter returns all items in the array that pass the set criteria

     case 'EDIT_EXPENSE': 
     return state.map((expense)=>{
       if(expense.id === action.id){
          return {
            ...expense,
            ...action.updates
          }
       } else{
          return expense
       }
     });
    default: 
      return state
  }
};