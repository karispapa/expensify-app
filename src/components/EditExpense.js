import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import {startEditExpense, startRemoveExpense} from '../actions/expenses'

export class EditExpense extends React.Component {
  // console.log(props) 
  
  onRemove = ()=>{
    this.props.startRemoveExpense(this.props.expense.id)
      this.props.history.push('/dashboard')
  };


  onSubmit = (expense)=>{
    this.props.startEditExpense(this.props.expense.id, expense)
    this.props.history.push('/dashboard')
  };

  render (){
  return (
  <div>
    <div className="page-header page-header__small">
      <div className="content-container" >
        <h1 className="page-header__title">Edit Expense</h1>
      </div>  
    </div>   
    <div className="content-container">
      <ExpenseForm 
        expense={this.props.expense}
        onSubmit={this.onSubmit}
      />
      <button className="button__remove" onClick={this.onRemove}>Remove Expense</button>
    </div>
  </div>
)}};
const mapStateToProps = (state, props)=> {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
};

const mapDispatchToProps = (dispatch, props)=>{
  return{
    startEditExpense: (id, updates)=> dispatch(startEditExpense(id, updates)),
    startRemoveExpense: (id)=> dispatch(startRemoveExpense({id}))
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);
