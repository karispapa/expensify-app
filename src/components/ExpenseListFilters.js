import React from 'react';
import { connect } from 'react-redux';
import { addTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates'
import 'react-dates/initialize';


export class ExpenseListFilters extends React.Component {
  state = {
    calenderFocused: null
  }

  onDatesChange =({startDate, endDate})=>{
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate)
  };

  onFocusChange = (calenderFocused)=>{
    this.setState(()=>({calenderFocused}))
  };

  onTextChange = (e)=>{
    this.props.addTextFilter(e.target.value)
  };

  onSortChange = (e)=>{
    if(e.target.value === "date"){
      this.props.sortByDate()
    }else if(e.target.value === "amount"){
      this.props.sortByAmount()
    }
};

  render(){
    return (
      <div className="content-container">
        <div className="input-group">
        <div className="input-group__item">
          <input type='text'
            className="text-input"  
            placeholder="Search Expenses"        
            value={this.props.filters.text} 
            onChange={this.onTextChange}/>  
        </div>
        <div className="input-group__item">
          <select 
            className="select-input"
            value={this.props.filters.sortBy} // controlled input - input controlled by js
              onChange={this.onSortChange}>
            <option value="date">Date</option>
            <option value="amount">Amount</option>
          </select>
        </div>
        <div className="input-group__item">
          <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          onFocusChange={this.onFocusChange}
          focusedInput={this.state.calenderFocused}
          numberOfMonths={1}
          isOutsideRange={()=> false}
          showClearDates={true}
          startDateId='id1'
          endDateId='id2'
          />
        </div>
        </div> 
        </div>
    )
  }

}

const mapStateToProps = (state)=>({ filters: state.filters })

const mapDispatchToProps = (dispatch)=>({
  addTextFilter: (text)=> dispatch(addTextFilter(text)), 
  sortByDate: ()=> dispatch(sortByDate()), 
  sortByAmount: ()=> dispatch(sortByAmount()), 
  setStartDate: (startDate)=> dispatch(setStartDate(startDate)), 
  setEndDate: (endDate)=> dispatch(setEndDate(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);