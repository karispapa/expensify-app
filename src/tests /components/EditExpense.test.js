
import React from 'react';
import { shallow } from 'enzyme';
import {EditExpense} from '../../components/EditExpense';
import expenses from '../fixtures/expenses';

let  wrapper, editExpense, startRemoveExpense, history

beforeEach(()=>{
  editExpense = jest.fn(); 
  startRemoveExpense = jest.fn();
  history = {push: jest.fn()}
  wrapper = shallow(<EditExpense expense={expenses[2]} editExpense={editExpense} 
    startRemoveExpense={startRemoveExpense} history={history}/>)
});

test('Should render Edit Expense page correctly', ()=>{
  expect(wrapper).toMatchSnapshot();
});

test('Should Handle Edit Expense', ()=>{
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
  expect(history.push).toHaveBeenLastCalledWith('/')
  expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2])
});


test('Should Handle Remove Expemse', ()=>{
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startRemoveExpense).toHaveBeenLastCalledWith(expenses[2].id)
})
