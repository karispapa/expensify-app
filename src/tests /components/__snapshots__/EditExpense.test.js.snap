// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Edit Expense page correctly 1`] = `
<div>
  <div
    className="page-header page-header__small"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 5000,
          "createdAt": 259200000,
          "description": "Coffee",
          "id": "3",
          "note": "Coffee",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button__remove"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
